torch/utils/hooks.py

```diff
@@ -116,6 +123,7 @@ def hook(grad_input, _):
                                    "got {}, but expected {}".format(len(res), len(grad_input)))
             return self._unpack_none(self.input_tensors_index, res)
         grad_fn.register_hook(hook)
+        self.counter += 1
```

Comment:
Why not move the for loop over the user hooks into `_set_user_hook` directly. That way it will always be called only once and we don't need to keep this counter around?

Reply:
Good idea, done!